# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFrameLoaderOwner.cpp
# Commit: acafbde0a546
# Full Hash: acafbde0a54602fcd1a634ddd772d666f7e96455
# Author: Olli Pettay <Olli.Pettay@helsinki.fi>
# Date: 2021-03-09 16:11:38
# Regressor Bug: 1696266
# File Overlap Count: 2
# Description:
#   Bug 1696266, limit the load types which may cause the page to enter bfcache, r=peterv
#   
#   The change to dom/base/nsFrameLoaderOwner.cpp is to log about the issues but still ensure we don't crash.
#   
#   I'd prefer to not put error loads to bfcache.
# ==============================================================================

diff -r f0d286830ebb -r acafbde0a546 dom/base/nsFrameLoaderOwner.cpp
--- a/dom/base/nsFrameLoaderOwner.cpp	Tue Mar 09 13:34:51 2021 +0000
+++ b/dom/base/nsFrameLoaderOwner.cpp	Tue Mar 09 13:42:43 2021 +0000
@@ -142,6 +142,18 @@
                 ("nsFrameLoaderOwner::ChangeRemotenessCommon: store the old "
                  "page in bfcache"));
             Unused << bc->SetIsInBFCache(true);
+            if (she->GetFrameLoader()) {
+              MOZ_LOG(gSHIPBFCacheLog, LogLevel::Debug,
+                      ("nsFrameLoaderOwner::ChangeRemotenessCommon: active "
+                       "session history entry "
+                       "has already an nsFrameLoader!, FIXME"));
+              MOZ_DIAGNOSTIC_ASSERT(false,
+                                    "The current session history entry has "
+                                    "already an nsFrameLoader!");
+              RefPtr<nsFrameLoader> oldFrameLoader = she->GetFrameLoader();
+              she->SetFrameLoader(nullptr);
+              oldFrameLoader->Destroy();
+            }
             she->SetFrameLoader(mFrameLoader);
             // Session history owns now the frameloader.
             mFrameLoader = nullptr;